Composer/packages/lib/indexers/src/validations/expressionValidation/utils.ts

```diff
@@ -41,22 +41,54 @@ export function findRequiredProperties(schema: any): { [key: string]: boolean }
   return required;
 }
 
-export function findTypes(schema: any): number[] {
+/*
+  case:
+    json filed type is string;
+    the $role is expression;
+    there is no oneOf;
+
+    we don't know which return type is correct,
+    so we think the expression support all return types.
+
+  case:
+    the $role is expression;
+    the oneOf has some normal types and there is no expression in oneOf
+
+    the return type is one of the types in oneOf filed.
+
+  case:
+    the $role is expression;
+    the oneOf has some normal types and there is a expression role in oneOf
+
+    the filed maybe a string expression and the expression return type is one of the types in oneOf filed besides the type with expression role.
+*/
+export function findTypes(schema: JSONSchema7): number[] {
```

Comment:
can you make this an arrow function, and great explanation above, but can you add a jsdoc style comment right above the function explaining the input and output? Currently I have no idea why is the function is returning an array of numbers and what these numbers represent. (can type ```number``` be replaced by ```StringMapExpressionType```)

Reply:
I do understand what it does, adding comments helps other devs coming across this code understand it quicker too. :)